File: src/NnCase.Importer/TFLite/LeakyRelu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics.Tensors;
using System.Text;
using NnCase.IR;
using NnCase.IR.Operators;

namespace NnCase.Importer
{
    /// <summary>
    /// TFLite importer LeakyRelu ops lowering.
    /// </summary>
    public partial class TFLiteImporter
    {
        private void ConvertLeakyRelu(tflite.Operator op)
        {
            var input = GetTensor(op.Inputs(0));
            var options = op.BuiltinOptions<tflite.LeakyReluOptions>().Value;

            var slope = _graph.AddNode(new Constant(DataType.Float32, BitConverter.GetBytes(options.Alpha), new Shape(1)));
            var mul = _graph.AddNode(new Binary(BinaryOperator.Mul, GetShape(input), slope.Shape, ValueRanges.DefaultFloat));
            var max = _graph.AddNode(new Binary(BinaryOperator.Max, GetShape(input), mul.Output.Shape, ValueRanges.DefaultFloat));
            mul.InputB.Connect(slope.Output);
            max.InputB.Connect(mul.Output);

            _inputTensors.Add(mul.InputA, op.Inputs(0));
            _inputTensors.Add(max.InputA, op.Inputs(0));
            _outputTensors.Add(op.Outputs(0), max.Output);
        }
    }
}

File: src/importer/tflite/ops/leaky_relu.cpp
#include "../tflite_importer.h"
#include <ir/ops/binary.h>
#include <ir/ops/constant.h>
#include <ir/ops/reduce.h>
#include <ir/ops/unary.h>

using namespace nncase;
using namespace nncase::importer;
using namespace nncase::ir;

DEFINE_TFLITE_LOWER(LEAKY_RELU)
{
    auto &input = get_tensor(op.inputs(), 0);
    auto &options = *op.builtin_options_as_LeakyReluOptions();
    auto in_shape = get_shape(*input.shape());

    auto alpha = graph_.emplace<constant>(options.alpha());
    auto mul = graph_.emplace<binary>(binary_mul, in_shape, alpha->output().shape(), value_range<float>::full());
    auto max = graph_.emplace<binary>(binary_max, in_shape, mul->output().shape(), value_range<float>::full());

    mul->input_b().connect(alpha->output());
    max->input_b().connect(mul->output());

    input_tensors_.emplace(&mul->input_a(), op.inputs()->Get(0));
    input_tensors_.emplace(&max->input_a(), op.inputs()->Get(0));
    output_tensors_.emplace(op.outputs()->Get(0), &max->output());
}
